File: services/control-plane/src/sync-tasks/utils/task-logger.ts
import { createPgPool } from '../../pg-connection.ts';
import { config } from '../../config.ts';
import { logInfo, logError } from '../../logger.ts';

// 创建单例数据库连接
let poolInstance: any = null;
function getPool() {
  if (!poolInstance) {
    poolInstance = createPgPool(config.databaseUrl, 'sync-tasks');
  }
  return poolInstance;
}

export type TaskStatus = 'running' | 'success' | 'failed';

export interface TaskLogRecord {
  id?: number;
  taskName: string;
  status: TaskStatus;
  startTime: Date;
  endTime?: Date;
  errorMessage?: string;
  syncCount?: number;
  dataSource?: string;
}

/**
 * 记录任务开始日志
 */
export async function logTaskStart(taskName: string): Promise<number> {
  const pool = getPool();
  const result = await pool.query(
    `INSERT INTO sync_task_logs (task_name, status, start_time) VALUES ($1, 'running', $2) RETURNING id`,
    [taskName, new Date()]
  );
  const taskId = result.rows[0].id;
  logInfo(`[sync-task] ${taskName} started, log id: ${taskId}`);
  return taskId;
}

/**
 * 记录任务成功日志
 */
export async function logTaskSuccess(
  taskId: number,
  syncCount: number = 0,
  dataSource: string = ''
): Promise<void> {
  const pool = getPool();
  await pool.query(
    `UPDATE sync_task_logs SET status = 'success', end_time = $1, sync_count = $2, data_source = $3 WHERE id = $4`,
    [new Date(), syncCount, dataSource, taskId]
  );
  logInfo(`[sync-task] task ${taskId} succeeded, sync count: ${syncCount}, data source: ${dataSource}`);
}

/**
 * 记录任务失败日志
 */
export async function logTaskFailed(
  taskId: number,
  errorMessage: string,
  syncCount: number = 0
): Promise<void> {
  const pool = getPool();
  await pool.query(
    `UPDATE sync_task_logs SET status = 'failed', end_time = $1, sync_count = $2, error_message = $3 WHERE id = $4`,
    [new Date(), syncCount, errorMessage, taskId]
  );
  logError(`[sync-task] task ${taskId} failed: ${errorMessage}`);
}

File: services/data-connector/src/utils/task_logger.py
from datetime import datetime
from typing import Optional
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from src.db.session import get_db
from src.models.sync_task_log import SyncTaskLog
from src.utils.logger import logger


class TaskLogger:
    """任务日志记录器，与control-plane格式完全兼容"""

    def __init__(self, db: Optional[AsyncSession] = None):
        self.db = db or next(get_db())

    async def log_task_start(self, task_name: str) -> int:
        """
        记录任务开始
        :param task_name: 任务名称
        :return: 任务日志ID
        """
        try:
            task_log = SyncTaskLog(
                task_name=task_name,
                status='running',
                start_time=datetime.utcnow()
            )
            self.db.add(task_log)
            await self.db.commit()
            await self.db.refresh(task_log)

            logger.info(f"[sync-task] {task_name} started, log id: {task_log.id}")
            return task_log.id

        except Exception as e:
            await self.db.rollback()
            logger.error(f"记录任务开始日志失败: {e}")
            raise

    async def log_task_success(self, task_id: int, sync_count: int = 0, data_source: str = '') -> None:
        """
        记录任务成功
        :param task_id: 任务日志ID
        :param sync_count: 同步记录数量
        :param data_source: 数据源名称
        """
        try:
            stmt = (
                update(SyncTaskLog)
                .where(SyncTaskLog.id == task_id)
                .values(
                    status='success',
                    end_time=datetime.utcnow(),
                    sync_count=sync_count,
                    data_source=data_source
                )
            )
            await self.db.execute(stmt)
            await self.db.commit()

            logger.info(f"[sync-task] task {task_id} succeeded, sync count: {sync_count}, data source: {data_source}")

        except Exception as e:
            await self.db.rollback()
            logger.error(f"记录任务成功日志失败: {e}")
            raise

    async def log_task_failed(self, task_id: int, error_message: str, sync_count: int = 0) -> None:
        """
        记录任务失败
        :param task_id: 任务日志ID
        :param error_message: 错误信息
        :param sync_count: 同步记录数量
        """
        try:
            stmt = (
                update(SyncTaskLog)
                .where(SyncTaskLog.id == task_id)
                .values(
                    status='failed',
                    end_time=datetime.utcnow(),
                    sync_count=sync_count,
                    error_message=error_message
                )
            )
            await self.db.execute(stmt)
            await self.db.commit()

            logger.error(f"[sync-task] task {task_id} failed: {error_message}")

        except Exception as e:
            await self.db.rollback()
            logger.error(f"记录任务失败日志失败: {e}")
            raise


# 全局单例实例
_task_logger_instance: Optional[TaskLogger] = None


def get_task_logger() -> TaskLogger:
    """获取任务日志记录器单例"""
    global _task_logger_instance
    if _task_logger_instance is None:
        _task_logger_instance = TaskLogger()
    return _task_logger_instance


# 导出便捷函数
async def log_task_start(task_name: str) -> int:
    return await get_task_logger().log_task_start(task_name)


async def log_task_success(task_id: int, sync_count: int = 0, data_source: str = '') -> None:
    await get_task_logger().log_task_success(task_id, sync_count, data_source)


async def log_task_failed(task_id: int, error_message: str, sync_count: int = 0) -> None:
    await get_task_logger().log_task_failed(task_id, error_message, sync_count)
